driver-core/src/main/java/com/datastax/driver/core/policies/ErrorAwarePolicy.java

```diff
@@ -0,0 +1,401 @@
+/*
+ *      Copyright (C) 2012-2015 DataStax Inc.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+package com.datastax.driver.core.policies;
+
+import com.datastax.driver.core.Cluster;
+import com.datastax.driver.core.ErrorTracker;
+import com.datastax.driver.core.Host;
+import com.datastax.driver.core.HostDistance;
+import com.datastax.driver.core.Statement;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.AbstractIterator;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+import java.util.ArrayDeque;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Chainable load balancing policy that will filter out hosts based on their errors rates.
+ * <p/>
+ * The policy bases its list of hosts on a child policy on which it will compute whether this host needs to included
+ * or excluded of the query plan.
+ * <p/>
+ * The policy uses a {@link ErrorAwarePolicyTracker} which is an extension of the base {@link ErrorTracker} that it
+ * will register automatically on the {@link Cluster}. The {@link ErrorAwarePolicyTracker} will be regularly updated
+ * and the policy will refer to it to take a decision on whether to include the child policy query plan's next host
+ * in the query plan. The policy will schedule a {@link ErrorTrackerUpdater} task on a fixed period that will call
+ * on each invocation {@link ErrorAwarePolicyTracker#update()}.
+ * <p/>
+ * The policy follows the builder pattern to be created, the {@link Builder} class can be created with
+ * {@link #builder} method.
+ */
+public class ErrorAwarePolicy implements ChainableLoadBalancingPolicy {
+
+    private final LoadBalancingPolicy childPolicy;
+
+    private final long updateRate;
+
+    private final static String EXECUTOR_SERVICE_NAME = "ErrorTracker updater task";
+
+    private final ScheduledExecutorService updaterService = Executors.newSingleThreadScheduledExecutor(new ThreadFactoryBuilder().setNameFormat(EXECUTOR_SERVICE_NAME).build());
```

Comment:
There are a few things that made me avoid being based on `currentTimeMillis` and preferred a fixed scheduled executor. I may point a few scenarios that come in my mind that wouldn't work in the algorithm you suggest (consider `updatePeriod = 200ms`) :

1) between t = 0 and t = 100ms we receive 20 errors from node1, then nothing anymore -> errors are never detected since the detection logic is driven by the errors happening
2) same scenario than above, except another error is happening at t = 210ms. We detect the errors and `startTime = now`. The startTime has relatively drifted of 10ms, the next errors would potentially be detected _starting_ at t = 410 ms, not before, and the first errors have been detected at t = 210ms.

We could try to correct those drifting and detection things with maths and averages and things. However, that would involve a lot of complications as of now, for a benefit **I don't manage to see**.

> currentTimeMillis should be good enough, we don't need to be very precise

The problem imo if we start being based on `currentTimeMillis` is that we're never going to be precise enough, plus if we can avoid checking for `currentTimeMillis` on every error...

Calculating and detecting the errorsRate and threshold in the main `Cluster` threads _may impact performances_ of the driver. I would rather go for a solution that is not degrading indeterministically the performances of the driver when facing big clusters with a lot of nodes.